SCClassLibrary/Common/Control/ServerNodeTree.sc

```diff
@@ -0,0 +1,198 @@
+ServerNodeTree {
+
+	var server;
+	var updateFunc, updater, <>refreshRate = 0.2;
+	var currentMsg, currentNodeTree = nil;
+	var <running = false;
+	var oscResp;
+	
+	*new { |server|
+		^super.new.init(server)
+	}
+	
+	init { |parentServer|
+		server = parentServer;
+	}
+
+	getNodeTree {
+		if(running) {
+			if(currentNodeTree.isNil) {
+				this.prFormatNodeTree;
+			};
+			^currentNodeTree
+		} {
+			^nil
+		};
+	}
+
+	run {
+		if(running.not) {
+			// Init osc receiver
+			oscResp = OSCFunc(
+				{ |msg|
+					// Only catch relevant messages
+					// (root node and controls enabled)
+					if((msg[1] == 1) and: { msg[2] == 0 }) {
+						this.prClearNodeTree;
+						currentMsg = msg;
+					};
+				},
+				'/g_queryTree.reply',
+				server.addr
+			).fix;
+			// Start auto update thread
+			updateFunc = {
+				updater = fork {
+					loop {
+						server.sendMsg(
+							"/g_queryTree",
+							0,
+							1
+						);
+						refreshRate.wait;
+					};
+				};
+				running = true;
+			};
+		};
+		updateFunc.value;
+		CmdPeriod.remove(updateFunc);
+	}
+
+	stop {
+		if(running) {
+			oscResp.free;
+			updater.stop;
+			CmdPeriod.remove(updateFunc);
+			running = false;
+		};
+	}
+
+	/*    PRIVATE METHODS    */
+
+	prClearNodeTree {
+		if(currentNodeTree.notNil) {
+			currentNodeTree.do({ |node|
+				this.prClearNodeDict(node);
+			});
+			currentNodeTree = nil;
+		};
+	}
+
+	// This allows to clear a node dict recursively
+	prClearNodeDict { |node|
+		if(node[\type] == \synth) {
+			node[\controls].clear;
+			node.clear;
+		} {
+			node[\children].do({ |child|
+				this.prClearNodeDict(child);
+			});
+			node[\children].clear;
+			node.clear;
+		};
+	}
+
+	// Called when all children in group have been processed
+	prFinalyzeGroup { |parentList, index|
+		var group = parentList.last;
+		group.removeAt(\nChildren);
+		parentList.pop;
+		// Check if this node is last child in parent,
+		// and finalyze parent if needed
+		if(parentList.size > 0) {
+			if(parentList.last[\children].size == parentList.last[\nChildren]) {
+				index = this.prFinalyzeGroup(parentList, index);
+			};
+		};
+		^(index + 2)
+	}
+
+	// Convert integer array representation of a node into readable format
+	prAnalyzeNode { |msg, index, nodes, parentList|
+		var node = Dictionary();
```

Comment:
If the names of keys are all symbols, it is *much* more efficient to use an `IdentityDictionary`

Reply:
Fixed.